bundle/edu.gemini.lchquery.servlet/src/main/scala/edu/gemini/lchquery/servlet/LchQueryServlet.scala

```diff
@@ -0,0 +1,147 @@
+      } yield (p,v)
+
+    val out: BufferedOutputStream = new BufferedOutputStream(response.getOutputStream)
+
+    // Handle IllegalArgumentException.
+    def illegalArgument(ex: IllegalArgumentException) = {
+      response.setStatus(LchQueryServlet.HttpResponseCodes.InvalidRequest.code)
+      response.setContentType("text/plain")
+      out.write(s"ERROR: ${ex.getMessage}".getBytes)
+    }
+
+    // Handle other Throwables that may arise.
+    def otherThrowable(t: Throwable, status: LchQueryServlet.HttpResponseCode): Unit = {
+      LchQueryServlet.Log.log(Level.WARNING, "could not process request", t)
+      response.setStatus(status.code)
+      response.setContentType("text/plain")
+      out.write(s"ERROR: ${t.getClass.getName}\n\n".getBytes)
+      Option(t.getMessage).foreach(s => out.write(s"$s\n\n".getBytes))
+
+      val writer = new PrintWriter(out)
+      t.printStackTrace(writer)
+      writer.close()
+    }
+
+
+    val pathInfo = request.getPathInfo
+    LchQueryFunctor.queryType(pathInfo) match {
+      case Some(queryType) =>
+        val programParams     = extractParams(LchQueryParam.ProgramParams)
+        val observationParams = extractParams(LchQueryParam.ObservationParams)
+
+        Try {
+          // Check that no illegal params have been supplied.
+          for {
+            p <- request.getParameterNames.asScala.map(_.asInstanceOf[String])
+            if !LchQueryParam.validParamName(p)
+          } throw new IllegalArgumentException(s"Invalid parameter: $p")
+
+          // Handle the query.
+          response.setStatus(LchQueryServlet.HttpResponseCodes.AllOK.code)
+          response.setContentType("application/xml")
+
+          odb.getQueryRunner(user.asJava).
+            queryPrograms(new LchQueryFunctor(queryType, programParams, observationParams)).
+            getResult.toXml
+
+        } recover {
+          case ex: IllegalArgumentException           => illegalArgument(ex)
+          case ex@(_:RemoteException | _:IOException) => otherThrowable(ex, LchQueryServlet.HttpResponseCodes.ServerError)
+          case ex                                     => otherThrowable(ex, LchQueryServlet.HttpResponseCodes.InvalidRequest)
+        }
+
+      case None =>
+        response.setStatus(LchQueryServlet.HttpResponseCodes.InvalidRequest.code)
+        response.setContentType("text/plain")
+        out.write(s"ERROR: Invalid query selector: $pathInfo. Should start with /programs, /observations or /targets".getBytes)
+    }
+
+    out.close()
+  }
+}
+
+object LchQueryServlet {
+  val Log = Logger.getLogger(LchQueryServlet.getClass.getName)
+
+  sealed abstract class HttpResponseCode(val code: Int)
+  object HttpResponseCodes {
+    case object AllOK extends HttpResponseCode(200)
+    case object InvalidRequest extends HttpResponseCode(400)
+    case object ServerError extends HttpResponseCode(500)
+  }
+
+  def apply(odb: IDBDatabaseService, user: java.util.Set[Principal]): LchQueryServlet =
```

Comment:
I guess we don't need this anymore. The bundle `Activator` was previously using a constructor with this signature, but we can just `.toSet` it with `scala.collection.JavaConversions` and be done with it. Removing it and updating the `Activtor`.